cmd/dra-example-kubeletplugin/state.go

```diff
@@ -73,113 +92,130 @@ func NewDeviceState(config *Config) (*DeviceState, error) {
+		allocatable:       allocatable,
+		checkpointManager: checkpointManager,
 	}
 
-	err = state.syncPreparedDevicesFromCRDSpec(&config.nascr.Spec)
+	checkpoints, err := state.checkpointManager.ListCheckpoints()
 	if err != nil {
-		return nil, fmt.Errorf("unable to sync prepared devices from CRD: %v", err)
+		return nil, fmt.Errorf("unable to list checkpoints: %v", err)
+	}
+
+	for _, c := range checkpoints {
+		if c == DriverPluginCheckpointFile {
+			return state, nil
+		}
+	}
+
+	checkpoint := newCheckpoint()
+	if err := state.checkpointManager.CreateCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
+		return nil, fmt.Errorf("unable to sync to checkpoint: %v", err)
 	}
 
 	return state, nil
 }
 
-func (s *DeviceState) Prepare(claimUID string, allocation nascrd.AllocatedDevices) ([]string, error) {
+func (s *DeviceState) Prepare(claimUID string, allocation AllocatedDevices) ([]string, error) {
 	s.Lock()
 	defer s.Unlock()
 
-	if s.prepared[claimUID] != nil {
-		cdiDevices, err := s.cdi.GetClaimDevices(claimUID, s.prepared[claimUID])
+	checkpoint := newCheckpoint()
+	if err := s.checkpointManager.GetCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
+		return nil, fmt.Errorf("unable to sync from checkpoint: %v", err)
+	}
+	preparedClaims := checkpoint.V1.PreparedClaims
+
+	if preparedClaims[claimUID] != nil {
+		cdiDevices, err := s.cdi.GetClaimDevices(claimUID, preparedClaims[claimUID])
 		if err != nil {
 			return nil, fmt.Errorf("unable to get CDI devices names: %v", err)
 		}
 		return cdiDevices, nil
 	}
 
-	prepared := &PreparedDevices{}
+	preparedDevices := &PreparedDevices{}
 
 	var err error
 	switch allocation.Type() {
-	case nascrd.GpuDeviceType:
-		prepared.Gpu, err = s.prepareGpus(claimUID, allocation.Gpu)
+	case gpucrd.GpuDeviceType:
+		preparedDevices.Gpu, err = s.prepareGpus(claimUID, allocation.Gpu)
 	default:
 		err = fmt.Errorf("unknown device type: %v", allocation.Type())
 	}
 	if err != nil {
-		return nil, fmt.Errorf("allocation failed: %v", err)
+		return nil, fmt.Errorf("praparation failed: %v", err)
 	}
 
-	err = s.cdi.CreateClaimSpecFile(claimUID, prepared)
+	err = s.cdi.CreateClaimSpecFile(claimUID, preparedDevices)
 	if err != nil {
 		return nil, fmt.Errorf("unable to create CDI spec file for claim: %v", err)
 	}
 
-	s.prepared[claimUID] = prepared
-
-	cdiDevices, err := s.cdi.GetClaimDevices(claimUID, s.prepared[claimUID])
+	cdiDevices, err := s.cdi.GetClaimDevices(claimUID, preparedDevices)
 	if err != nil {
 		return nil, fmt.Errorf("unable to get CDI devices names: %v", err)
 	}
+
+	preparedClaims[claimUID] = preparedDevices
+	if err := s.checkpointManager.CreateCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
+		return nil, fmt.Errorf("unable to sync to checkpoint: %v", err)
+	}
+
 	return cdiDevices, nil
 }
 
 func (s *DeviceState) Unprepare(claimUID string) error {
 	s.Lock()
 	defer s.Unlock()
 
-	if s.prepared[claimUID] == nil {
+	checkpoint := newCheckpoint()
```

Comment:
Why does the checkpoint need to be loaded every time unprepare or prepare is called? Is it some sort of protection from kubelet-plugin Pod being offline / suspended temporarily? If not, my impression was that the checkpoint can be loaded just once when the kubelet-plugin object is being created and new state object is created. After state is ready, then the prepared claims need to be saved into the checkpoint every time `Prepare` or `Unprepare` is called trusting that State object is the source of truth, and ignoring old data in snapshots.